internal/controller/clusterdeployment_controller.go

```diff
@@ -665,85 +677,96 @@ func (r *ClusterDeploymentReconciler) Delete(ctx context.Context, cd *kcm.Cluste
 		}
 	}()
 
-	hr := &hcv2.HelmRelease{}
+	// Without explicitly deleting the Profile object, we run into a race condition
+	// which prevents Sveltos objects from being removed from the management cluster.
+	// It is detailed in https://github.com/projectsveltos/addon-controller/issues/732.
+	// We may try to remove the explicit call to Delete once a fix for it has been merged.
+	// TODO(https://github.com/K0rdent/kcm/issues/526).
+	if err := sveltos.DeleteProfile(ctx, r.Client, cd.Namespace, cd.Name); err != nil {
+		return ctrl.Result{}, err
+	}
 
-	if err := r.Client.Get(ctx, client.ObjectKeyFromObject(cd), hr); err != nil {
-		if !apierrors.IsNotFound(err) {
+	err = r.Client.Get(ctx, client.ObjectKeyFromObject(cd), &hcv2.HelmRelease{})
+	if err == nil { // if NO error
+		if err := helm.DeleteHelmRelease(ctx, r.Client, cd.Name, cd.Namespace); err != nil {
 			return ctrl.Result{}, err
 		}
-		cluster := &metav1.PartialObjectMetadata{}
-		cluster.SetGroupVersionKind(schema.GroupVersionKind{
-			Group:   "cluster.x-k8s.io",
-			Version: "v1beta1",
-			Kind:    "Cluster",
-		})
-		if err = r.Client.Get(ctx, client.ObjectKeyFromObject(cd), cluster); apierrors.IsNotFound(err) {
-			l.Info("Removing Finalizer", "finalizer", kcm.ClusterDeploymentFinalizer)
-			if controllerutil.RemoveFinalizer(cd, kcm.ClusterDeploymentFinalizer) {
-				if err := r.Client.Update(ctx, cd); err != nil {
-					return ctrl.Result{}, fmt.Errorf("failed to update clusterDeployment %s/%s: %w", cd.Namespace, cd.Name, err)
-				}
-			}
-			l.Info("ClusterDeployment deleted")
+
+		l.Info("HelmRelease still exists, retrying")
```

Comment:
With the current implementation, the kcm-controller will not continue the deletion reconciliation when the HelmRelease still exists. HelmRelease will not be deleted until either the timeout is reached or all the objects are deleted. The `provider cluster` will not be deleted until the `releaseProviderCluster` function is called. So, here is a circular dependency:
```
HelmRelease is not deleted until CAPI Cluster exists <-> CAPI Cluster will not be released until HelmRelease is gone
```
and the deletion will be stuck until the HelmRelease is gone due to timeout.

Reply:
So that is the previous implementation and the current one (as of this PR), right? Because the previous one was performing the same loop depending on a `HelmRelease` object. Hence the question: what should be done first?

Reply:
Yes, the previous implementation wasn’t entirely correct, but the controller initially didn’t stop further reconciliation if the HelmRelease still existed. This part is mostly correct. But, when the HelmRelease still exists, we shouldn’t return early. Instead, we should continue with the next steps.

Reply:
I got it! Fixed. Thanks for pointing that out, I haven't been aware of such a dependency.

Reply:
There is one thing though that concerns me: if `Machine` objects still exist (CAPI) then we continue the execution but I think that we should not continue it and throw an error (probably a sentinel one) hence trying to release the provider's cluster at first, shouldn't we?

Reply:
No, I don't think we should throw an error: the controller will be stuck. The circular dependency still exists: now, the `releaseProviderCluster` is called before the HelmRelease deletion. But the Machines will never be triggered for the deletion until the HelmRelease is triggered for the deletion. And we will never get to see machines' deletion happen since it'll never be triggered for the deletion. So, we should continue executing the next steps.